File: 39/CombinationSum.cpp
#include<iostream>
#include<vector>
using namespace std;
void backtrack(vector<vector<int>>& res,vector<int>& arr,int ind,int currsum,int n,int target,vector<int>& nums)
{
    if(currsum==target)
    {
        res.push_back(arr);
        return;
    }
    if(currsum>target)
    {
        return;
    }
    for(int i=ind;i<n;i++)
    {
        arr.push_back(nums[i]);
        backtrack(res,arr,i,currsum+nums[i],n,target,nums);
        arr.pop_back();
    }
}
vector<vector<int>> combinationsum(vector<int> nums,int target)
{
    vector<int> arr;
    vector<vector<int>> res;
    backtrack(res,arr,0,0,nums.size(),target,nums);
    return res;
}
int main()
{
    vector<int> nums={2,3,6,7};
    int target=7;
    vector<vector<int>> res=combinationsum(nums,target);

    for(auto& r:res)
    {
        for(int i:r)
        {
            cout<<i<<",";
        }
        cout<<endl;
    }
    return 0;
}

File: 39/CombinationSum.py
from typing import List
class Solution:
    def combinationSum(self,nums,target):
        res=[]
        def backtrack(curr_sum,ind,arr:List[int]):
            if curr_sum==target:
                res.append(arr[:])
                return
            if curr_sum>target:
                return
            for i in range(ind,len(nums)):
                arr.append(nums[i])
                backtrack(curr_sum+nums[i],i,arr)
                arr.pop()
        backtrack(0,0,[])
        return res

nums=[2,5,6,9]
target=9
sol=Solution()
print(sol.combinationSum(nums,target))
